File: data-structure/Stack/ArrayStack.ts
import MyArray from '../Array/Array';

interface Stack<E> {
  getSize(): number;
  isEmpty(): boolean;
  push(e: E): void;
  pop(): E;
  peek(): E;
}

export default class ArrayStack<E> implements Stack<E> {
  private array: MyArray<E>;

  constructor(capacity: number = 10) {
    this.array = new MyArray<E>(capacity);
  }

  getCapacity(): number {
    return this.array.getCapacity();
  }

  getSize(): number {
    return this.array.getSize();
  }

  isEmpty(): boolean {
    return this.array.isEmpty();
  }

  push(e: E): void {
    this.array.addLast(e);
  }

  pop(): E {
    return this.array.removeLast();
  }

  peek(): E {
    return this.array.getLast();
  }

  toString(): string {
    let res = '';
    res += 'Stack: ';
    res += '[';
    for (let i = 0; i < this.getSize(); i++) {
      res += this.array.get(i);
      if (i !== this.array.getSize() - 1) {
        res += ', ';
      }
    }
    res += '] top';

    return res;
  }
}

// let s = new ArrayStack<number>();
// for (let i = 0; i < 5; i++) {
//   s.push(i);
//   console.log(s.toString());
// }
// s.pop();
// console.log(s.toString());

File: DataStructureJava/ArrayStack.java
public class ArrayStack<E> implements Stack<E> {
    Array<E> array;
    public ArrayStack(int capacity) {
        array = new Array<>(capacity);
    }

    public ArrayStack() {
        array = new Array<>();
    }

    // 时间复杂度O(1)
    // 获取栈中当前有效元素的个数
    @Override
    public int getSize() {
        return array.getSize();
    }

    // 时间复杂度O(1)
    // 判断当前栈是否为空
    @Override
    public boolean isEmpty() {
        return array.isEmpty();
    }

    // 时间复杂度O(1)
    // 获取当前栈的容量
    public int getCapacity() {
        return array.getCapacity();
    }

    // 时间复杂度O(1)
    // 向栈中添加一个元素
    @Override
    public void push(E e) {
        array.addLast(e);
    }

    // 时间复杂度O(1)
    // 出栈一个元素
    @Override
    public E pop() {
        return array.removeLast();
    }

    // 时间复杂度O(1)
    // 查看栈顶的元素
    @Override
    public E peek() {
        return array.getLast();
    }

    @Override
    public String toString() {
        StringBuilder res = new StringBuilder();
        res.append("Stack: [");
        for (int i = 0; i < array.getSize(); i++) {
            res.append(array.get(i));
            if (i != array.getSize() - 1)
                res.append(", ");
        }
        res.append("] top");
        return res.toString();
    }

    public static void main(String[] args) {
        ArrayStack<Integer> stack = new ArrayStack<>();
        for (int i = 0; i < 5; i++) {
            stack.push(i);
            System.out.println(stack);
        }

        stack.pop();
        System.out.println(stack);

        int topEl = stack.peek();
        System.out.println(topEl);
    }
}
